server/channels/app/file.go

```diff
@@ -1594,3 +1594,64 @@ func getFileExtFromMimeType(mimeType string) string {
 	}
 	return "jpg"
 }
+
+func (a *App) PermanentDeleteFilesByPost(c request.CTX, postID string) *model.AppError {
+	fileInfos, err := a.Srv().Store().FileInfo().GetForPost(postID, false, true, true)
+	if err != nil {
+		return model.NewAppError("PermanentDeleteFilesByPost", "app.file_info.get_by_post_id.app_error", nil, "", http.StatusInternalServerError).Wrap(err)
+	}
+	if len(fileInfos) == 0 {
+		c.Logger().Info("No files found for post", mlog.String("post_id", postID))
+		return model.NewAppError("PermanentDeleteFilesByPost", "app.file_info.get_by_post_id.app_error", nil, "", http.StatusNotFound)
+	}
+
+	a.RemoveFilesFromFileStore(c, fileInfos)
+
+	err = a.Srv().Store().FileInfo().PermanentDeleteForPost(c, postID)
+	if err != nil {
+		return model.NewAppError("PermanentDeleteFilesByPost", "app.file_info.permanent_delete_for_post.app_error", nil, "", http.StatusInternalServerError).Wrap(err)
+	}
+
+	a.Srv().Store().FileInfo().InvalidateFileInfosForPostCache(postID, true)
```

Comment:
I think that's in order to avoid a race condition due to `App.DeletePost` using a go routine for `deletePostFiles`

Reply:
If I understand the code correctly, there can be two entries in the `FileInfoCache` for the same post, one indexed with the `PostID` and the other indexed with `PostID_deleted`. The latter is used when the post is marked as deleted

I _think_ we are fine with not calling `InvalidateFileInfosForPostCache` as `DeletePost` should have cleared the cache entry index with `PostID`. But I wonder if we should still clear both cache entries just to be on the safe side. (And add explain the concerns via a comment)

Reply:
I took a look through our code and you're right, I think we need both. I'm confused why we do this with our cache keys because it doesn't make much sense but looks at this code snippet:
```go
func (s LocalCacheFileInfoStore) GetForPost(postId string, readFromMaster, includeDeleted, allowFromCache bool) ([]*model.FileInfo, error) {
	if !allowFromCache {
		return s.FileInfoStore.GetForPost(postId, readFromMaster, includeDeleted, allowFromCache)
	}

	cacheKey := postId
	if includeDeleted {
		cacheKey += "_deleted"
	}

	var fileInfo []*model.FileInfo
	if err := s.rootStore.doStandardReadCache(s.rootStore.fileInfoCache, cacheKey, &fileInfo); err == nil {
		return fileInfo, nil
	}

	fileInfos, err := s.FileInfoStore.GetForPost(postId, readFromMaster, includeDeleted, allowFromCache)
	if err != nil {
		return nil, err
	}

	if len(fileInfos) > 0 {
		s.rootStore.doStandardAddToCache(s.rootStore.fileInfoCache, cacheKey, fileInfos)
	}

	return fileInfos, nil
}
```
If someone includes the `include_deleted` url parameter it will append `_deleted` to the cache key, it won't find it in the cache layer and then it will query the db. The db will return the file, doesn't matter if the file is deleted or not because `include_deleted` means we want it either way. We then create a cache key with `_deleted` appended without actually checking if the file is deleted. If someone tries to fetch that file without the `include_deleted` query param it will skip the cache layer and query the db. We probably need to file a ticket to fix this and only append `_deleted` for files that are marked as deleted. I'm not even sure why we append `_deleted` tbh